### DISCUSSION TITLE
Multiple sites

### DISCUSSION THREAD

**Discussion by maralkumbe** [pimcore/pimcore#11118] [Q&A / Support]
[For "sites" I mean environments/projects for different customers]
Hi, sorry I'm new to Pimcore and I'm sure what I'm asking is written somewhere I can't find.
I set up a demo and a skeleton docker-compose based environments to try and my developer collegue says all works well :)
We are planning now to deploy different projects for different customers each with its own ssl certificate. Would a solution with one reverse proxy in front of one docker-compose network for each customer make sense or one pimcore instance (docker-compose) could manage all projects and it would also optimize the maintenance work on the long run (e.g. for updating).
In general any suggestion/docs about production deployment (especially for multiple instances if available) would be appreciated.
Thank you,
Marco

**Comment by maralkumbe** [comment]
Hi, I would like to share the same host 443 port with the different pimcore instances so I tried configuring two different docker-compose environments for two different pimcore instances on the same pimcore-server and setting up an https reverse proxy on another server.
So e.g. the reverse proxy on one server exposes the pimcore web app (http://<pimcore.server>:<customer.port>/) as  https://<www.example.com>/ 
The homepage is loaded correctly with https managed by the proxy, is there anything to set in pimcore to manage the fact it is working behind a reverse proxy or this is an usual deployment configuration that needs no other step on pimcore side?

Thank you

**Comment by zangab** [comment]
so u want to deploy 1 instance for - lets say - 4 different clients?
lol, they all share the same data u know. that's just unprofessional.

**Reply by maralkumbe** [reply]
Thank you zangab for your confirmation, I'm working on the sysadmin side and didn't know in fact, and didn't find the info about pimcore being designed to manage multiple instances on the same db/docker-compose environment or not. 

So I suppose that having one docker-compose environment for each instance/client as I did is the right choice. 

Now I cannot use one public IP address for each instance and this the main reason for I put an ssl proxy in front of them instead of using pimcore's ssl modules.

So to close the circle I would like to know if being behind an ssl proxy is transparent to pimcore that can remain working in plain http (each instance exposed on a different tcp port) or it requires some custom config?

**Reply by zangab** [reply]
i'm not that deep into docker but i suppose each project runs it its own container now?
running multiple sites on a server should be possible though, similar to apache's vhost.

**Reply by maralkumbe** [reply]
I also think it should be possible but I can't find any docs about it.

**Reply by zangab** [reply]
you would need a nginx as proxy, that runs on port 80 and then redirect those requests to the correct docker instance. that's how i'd try to solve it imho. this is not a pimcore issue.
something similar to this: https://www.digitalocean.com/community/tutorials/how-to-set-up-nginx-server-blocks-virtual-hosts-on-ubuntu-16-04

**Reply by maralkumbe** [reply]
You are right, an ssl nginx proxy is what I put in front of the different docker-compose networks and it works.

One particular problem I got was caused by PUT methods which where not enabled by default in the nginx proxy and are required to allow some uploads.

Another problem that can be managed, but I don't see a global option to do that, are mixed content produced if the pimcore instance is behind an ssl proxy: 
1. the browser makes an https request to the proxy 
2. which makes an http request to the tcp port exposed by the docker-compose network
3. pimcore receives an http request and builds a web page that contains http urls instead of https

Maybe there is a global option to say pimcore (or symfony?) you are behind an ssl proxy and you should build relative urls (that the browser keeps calling in https) or force an https schema.

**Reply by zangab** [reply]
just force ssl (rewrite https) thats pretty easy to do via nginx

**Reply by maralkumbe** [reply]
Thank you zangab, you mean enable ssl on each pimcore instance and configure the ssl proxy to reencrypt traffic this way: 

browser --ssl--> proxy --ssl--> pimcore 
?

I thought about exposing a different ssl custom port for each docker-compose network and if there is no other solution it could be ok.

I would prefer an easier solution that does not require to manage ssl certificate both on proxy and web server and avoid useless cpu load but it could be the simplest solution after all.
